File: web/src/admin-dsl/builder.ts
import type {
  AdminActionRef,
  AdminJsonObject,
  AdminJsonValue,
  AdminNode,
  AdminNodeKind,
  AdminPage,
  AdminQueryRef,
  AdminShellKind,
} from "./schema";

type NodeInput = AdminNode | AdminNodeBuilder;
type ActionInput = AdminActionRef | AdminActionBuilder;
type QueryInput = AdminQueryRef | QueryBuilder | string;

function clone<T>(value: T): T {
  return JSON.parse(JSON.stringify(value));
}

function unwrapNode(input: NodeInput): AdminNode {
  return input instanceof AdminNodeBuilder ? input.toJSON() : input;
}

function unwrapAction(input: ActionInput): AdminActionRef {
  return input instanceof AdminActionBuilder ? input.toJSON() : input;
}

function unwrapQuery(input: QueryInput): AdminQueryRef {
  if (typeof input === "string") return { id: input };
  return input instanceof QueryBuilder ? input.toJSON() : input;
}

export class AdminNodeBuilder<P extends AdminJsonObject = AdminJsonObject> {
  protected node: AdminNode<P>;

  constructor(kind: AdminNodeKind, props?: P) {
    this.node = { kind, props } as AdminNode<P>;
  }

  id(id: string) {
    this.node.meta = { ...(this.node.meta || {}), id };
    return this;
  }

  name(name: string) {
    this.node.meta = { ...(this.node.meta || {}), name };
    return this;
  }

  region(region: NonNullable<NonNullable<AdminNode["meta"]>["region"]>) {
    this.node.meta = { ...(this.node.meta || {}), region };
    return this;
  }

  section(dataSection: string) {
    this.node.meta = { ...(this.node.meta || {}), dataSection };
    return this;
  }

  part(dataPart: string) {
    this.node.meta = { ...(this.node.meta || {}), dataPart };
    return this;
  }

  note(note: string) {
    this.node.meta = { ...(this.node.meta || {}), note };
    return this;
  }

  props(props: AdminJsonObject) {
    this.node.props = { ...this.node.props, ...props } as P;
    return this;
  }

  children(...children: NodeInput[]) {
    this.node.children = children.map(unwrapNode);
    return this;
  }

  add(...children: NodeInput[]) {
    this.node.children = [...(this.node.children || []), ...children.map(unwrapNode)];
    return this;
  }

  action(slot: string, actionRef: ActionInput) {
    const current = (this.node.props?.actions && typeof this.node.props.actions === "object" && !Array.isArray(this.node.props.actions))
      ? this.node.props.actions as AdminJsonObject
      : {};
    this.node.props = {
      ...(this.node.props || {}),
      actions: { ...current, [slot]: unwrapAction(actionRef) as unknown as AdminJsonValue },
    } as unknown as P;
    return this;
  }

  actions(...actions: ActionInput[]) {
    this.node.props = {
      ...(this.node.props || {}),
      actions: actions.map(unwrapAction) as unknown as AdminJsonValue,
    } as unknown as P;
    return this;
  }

  query(queryRef: QueryInput, params?: AdminJsonObject) {
    const q = typeof queryRef === "string" ? { id: queryRef, params } : unwrapQuery(queryRef);
    this.node.props = { ...(this.node.props || {}), query: q as unknown as AdminJsonValue } as unknown as P;
    return this;
  }

  empty(emptyState: NodeInput) {
    this.node.props = { ...(this.node.props || {}), empty: unwrapNode(emptyState) as unknown as AdminJsonValue } as unknown as P;
    return this;
  }

  state(state: "idle" | "loading" | "empty" | "error" | "dirty" | "pending" | "success") {
    this.node.props = { ...(this.node.props || {}), state } as unknown as P;
    return this;
  }

  values(values: AdminJsonObject) {
    this.node.props = { ...(this.node.props || {}), values } as unknown as P;
    return this;
  }

  errors(errors: AdminJsonObject) {
    this.node.props = { ...(this.node.props || {}), errors } as unknown as P;
    return this;
  }

  submit(actionRef: ActionInput) {
    return this.action("submit", actionRef);
  }

  cancel(actionRef: ActionInput) {
    return this.action("cancel", actionRef);
  }

  dirty(dirty = true) {
    this.node.props = { ...(this.node.props || {}), dirty } as unknown as P;
    return this;
  }

  pending(pending = true) {
    this.node.props = { ...(this.node.props || {}), pending } as unknown as P;
    return this;
  }

  layoutPolicy(policy: AdminJsonObject) {
    this.node.props = { ...(this.node.props || {}), layoutPolicy: policy } as unknown as P;
    return this;
  }

  adaptive(views: AdminJsonObject) {
    this.node.props = { ...(this.node.props || {}), adaptive: views } as unknown as P;
    return this;
  }

  toJSON(): AdminNode<P> {
    return clone(this.node);
  }
}

export class AdminPageBuilder {
  private page: AdminPage;

  constructor(id: string, title: string, shell: AdminShellKind = "admin") {
    this.page = {
      schemaVersion: 1,
      id,
      title,
      shell: { kind: shell },
      nodes: [],
    };
  }

  title(title: string) {
    this.page.title = title;
    return this;
  }

  describe(description: string) {
    this.page.description = description;
    return this;
  }

  schemaVersion(schemaVersion: AdminPage["schemaVersion"]) {
    this.page.schemaVersion = schemaVersion;
    return this;
  }

  shell(kind: AdminShellKind, props: AdminJsonObject = {}) {
    this.page.shell = { kind, props };
    return this;
  }

  meta(meta: NonNullable<AdminPage["meta"]>) {
    this.page.meta = { ...(this.page.meta || {}), ...meta };
    return this;
  }

  content(...nodes: NodeInput[]) {
    this.page.nodes = nodes.map(unwrapNode);
    return this;
  }

  add(...nodes: NodeInput[]) {
    this.page.nodes.push(...nodes.map(unwrapNode));
    return this;
  }

  toolbar(...actions: ActionInput[]) {
    this.page.nodes.unshift(admin.toolbar(...actions).toJSON());
    return this;
  }

  modals(...nodes: NodeInput[]) {
    this.page.modals = nodes.map(unwrapNode);
    return this;
  }

  drawers(...nodes: NodeInput[]) {
    this.page.drawers = nodes.map(unwrapNode);
    return this;
  }

  toJSON(): AdminPage {
    return clone(this.page);
  }
}

export class AdminActionBuilder {
  constructor(private readonly ref: AdminActionRef) {}

  payload(payload: AdminJsonValue) {
    return new AdminActionBuilder({ ...this.ref, payload });
  }

  options(options: AdminJsonObject) {
    return new AdminActionBuilder({ ...this.ref, options: { ...(this.ref.options || {}), ...options } });
  }

  intent(intent: NonNullable<AdminActionRef["intent"]>) {
    return new AdminActionBuilder({ ...this.ref, intent });
  }

  priority(priority: NonNullable<AdminActionRef["priority"]>) {
    return new AdminActionBuilder({ ...this.ref, priority });
  }

  placement(placement: NonNullable<AdminActionRef["placement"]>) {
    return new AdminActionBuilder({ ...this.ref, placement });
  }

  presentation(presentation: NonNullable<AdminActionRef["presentation"]>) {
    return new AdminActionBuilder({ ...this.ref, presentation });
  }

  confirm(options: AdminJsonObject = {}) {
    return new AdminActionBuilder({
      ...this.ref,
      requiresConfirmation: true,
      options: { ...(this.ref.options || {}), confirmation: options },
    });
  }

  disabled(disabled = true) {
    return new AdminActionBuilder({ ...this.ref, disabled });
  }

  loading(loading = true) {
    return new AdminActionBuilder({ ...this.ref, loading });
  }

  accessibilityLabel(accessibilityLabel: string) {
    return new AdminActionBuilder({ ...this.ref, accessibilityLabel });
  }

  toJSON(): AdminActionRef {
    return clone(this.ref);
  }
}

export class QueryBuilder {
  constructor(private readonly ref: AdminQueryRef) {}

  params(params: AdminJsonObject) {
    return new QueryBuilder({ ...this.ref, params });
  }

  toJSON(): AdminQueryRef {
    return clone(this.ref);
  }
}

function makeAction(
  type: AdminActionRef["type"],
  target: string,
  label?: string,
  payload?: AdminJsonValue,
  options?: AdminJsonObject,
): AdminActionRef {
  return {
    type,
    target,
    ...(label === undefined ? {} : { label }),
    ...(payload === undefined ? {} : { payload }),
    ...(options === undefined ? {} : { options }),
  } as AdminActionRef;
}

export const action = {
  open: (target: string, label?: string, payload?: AdminJsonValue) => new AdminActionBuilder(makeAction("open", target, label, payload)),
  close: (target = "current") => new AdminActionBuilder(makeAction("close", target)),
  navigate: (target: string, label?: string, payload?: AdminJsonValue) => new AdminActionBuilder(makeAction("navigate", target, label, payload)),
  mutation: (target: string, label?: string, payload?: AdminJsonValue) => new AdminActionBuilder(makeAction("mutation", target, label, payload)),
  confirm: (target: string, label?: string, payload?: AdminJsonValue) => new AdminActionBuilder(makeAction("confirm", target, label, payload)).intent("danger").confirm(),
  refresh: (target: string, label = "Refresh", payload?: AdminJsonValue) => new AdminActionBuilder(makeAction("refresh", target, label, payload)),
  upload: (target: string, label?: string, options: AdminJsonObject = {}) => new AdminActionBuilder(makeAction("upload", target, label, undefined, options)),
  primary: (target: string, label?: string, payload?: AdminJsonValue) => new AdminActionBuilder(makeAction("mutation", target, label, payload)).intent("primary").priority("primary"),
  secondary: (target: string, label?: string, payload?: AdminJsonValue) => new AdminActionBuilder(makeAction("mutation", target, label, payload)).intent("neutral").priority("secondary"),
  danger: (target: string, label?: string, payload?: AdminJsonValue) => new AdminActionBuilder(makeAction("mutation", target, label, payload)).intent("danger").priority("secondary").confirm(),
  ghost: (target: string, label?: string, payload?: AdminJsonValue) => new AdminActionBuilder(makeAction("mutation", target, label, payload)).intent("neutral").priority("tertiary").presentation("link"),
};

export const query = {
  ref: (id: string, params?: AdminJsonObject) => new QueryBuilder({ id, ...(params === undefined ? {} : { params }) } as AdminQueryRef),
};

function node(kind: AdminNodeKind, props: AdminJsonObject = {}, ...children: NodeInput[]) {
  return new AdminNodeBuilder(kind, props).children(...children);
}

export const admin = {
  page: (id: string, title: string) => new AdminPageBuilder(id, title, "admin"),
  dashboard: (id = "dashboard", title = "Dashboard") => new AdminPageBuilder(id, title, "dashboard"),
  calendarPage: (id = "calendar", title = "Calendar") => new AdminPageBuilder(id, title, "calendar"),
  settings: (id: string, title: string) => new AdminPageBuilder(id, title, "settings"),

  pageHeader: (props: AdminJsonObject = {}) => node("pageHeader", props),
  dashboardGrid: (props: AdminJsonObject = {}, ...children: NodeInput[]) => node("dashboardGrid", props, ...children),
  section: (title: string, props: AdminJsonObject = {}, ...children: NodeInput[]) => node("section", { title, ...props }, ...children),
  toolbar: (...actions: ActionInput[]) => node("toolbar", { actions: actions.map((a) => unwrapAction(a)) as unknown as AdminJsonValue }),
  cardGrid: (...children: NodeInput[]) => node("cardGrid", {}, ...children),
  panel: (title: string, props: AdminJsonObject = {}, ...children: NodeInput[]) => node("panel", { title, ...props }, ...children),
  splitPane: (props: AdminJsonObject = {}, ...children: NodeInput[]) => node("splitPane", props, ...children),
  tabs: (tabs: AdminJsonObject[], value: string, props: AdminJsonObject = {}) => node("tabs", { tabs, value, ...props }),
  editableList: (id: string, items: AdminJsonObject[], props: AdminJsonObject = {}) => node("editableList", { id, items, ...props }).id(id),
  monthAvailabilityGrid: (id: string, days: AdminJsonObject[], props: AdminJsonObject = {}) => node("monthAvailabilityGrid", { id, days, ...props }).id(id),
  previewFrame: (id: string, props: AdminJsonObject = {}) => node("previewFrame", { id, ...props }).id(id),
  comparisonTable: (id: string, rows: AdminJsonObject[], props: AdminJsonObject = {}) => node("comparisonTable", { id, rows, ...props }).id(id),
  monthCalendar: (id: string, props: AdminJsonObject = {}) => node("monthCalendar", { id, ...props }).id(id),
  diffView: (id: string, changes: AdminJsonObject[], props: AdminJsonObject = {}) => node("diffView", { id, changes, ...props }).id(id),

  metric: (label: string, value: string | number, props: AdminJsonObject = {}) => node("metricCard", { label, value, ...props }),
  summary: (title: string, props: AdminJsonObject = {}) => node("summaryCard", { title, ...props }),
  badge: (label: string, tone = "neutral") => node("statusBadge", { label, tone }),
  emptyState: (title: string, props: AdminJsonObject = {}) => node("emptyState", { title, ...props }),
  inlineError: (title: string, props: AdminJsonObject = {}) => node("inlineError", { title, ...props }),
  markdown: (markdown: string, props: AdminJsonObject = {}) => node("markdownBlock", { markdown, ...props }),
  kvList: (items: AdminJsonObject[], props: AdminJsonObject = {}) => node("kvList", { items, ...props }),
  activityFeed: (items: AdminJsonObject[], props: AdminJsonObject = {}) => node("activityFeed", { items, ...props }),
  imageGrid: (items: AdminJsonObject[], props: AdminJsonObject = {}) => node("imageGrid", { items, ...props }),
  imageGallery: (id: string, images: AdminJsonObject[], props: AdminJsonObject = {}) => node("imageGallery", { id, images, ...props }).id(id),
  loadingState: (title: string, props: AdminJsonObject = {}) => node("loadingState", { title, ...props }),

  resourceList: (id: string, props: AdminJsonObject = {}, ...children: NodeInput[]) => node("resourceList", { id, ...props }, ...children),
  filterBar: (filters: AdminJsonObject[], value?: string, props: AdminJsonObject = {}) => node("filterBar", { filters, value: value || null, ...props }),
  searchBox: (placeholder = "Search", props: AdminJsonObject = {}) => node("searchBox", { placeholder, ...props }),

  form: (id: string, props: AdminJsonObject = {}, ...children: NodeInput[]) => node("form", { id, ...props }, ...children),
  fieldGroup: (title: string, ...children: NodeInput[]) => node("fieldGroup", { title }, ...children),
  saveBar: (props: AdminJsonObject = {}) => node("saveBar", props),

  calendarWeek: (id: string, props: AdminJsonObject = {}, ...children: NodeInput[]) => node("calendarWeek", { id, ...props }, ...children),
  appointmentBlock: (id: string, props: AdminJsonObject = {}) => node("appointmentBlock", { id, ...props }),
  availabilityBlock: (id: string, props: AdminJsonObject = {}) => node("availabilityBlock", { id, ...props }),
  timeOffBlock: (id: string, props: AdminJsonObject = {}) => node("timeOffBlock", { id, ...props }),

};

export const surface = {
  drawer: (id: string, props: AdminJsonObject = {}, ...children: NodeInput[]) => node("drawer", { id, presentation: "drawer", ...props }, ...children).region("drawer"),
  modal: (id: string, props: AdminJsonObject = {}, ...children: NodeInput[]) => node("modal", { id, presentation: "modal", ...props }, ...children).region("modal"),
  sheet: (id: string, props: AdminJsonObject = {}, ...children: NodeInput[]) => node("sheet", { id, presentation: "sheet", ...props }, ...children).region("drawer"),
  confirm: (id: string, props: AdminJsonObject = {}) => node("confirmDialog", { id, presentation: "confirm", ...props }).region("modal"),
  detailPanel: (id: string, props: AdminJsonObject = {}, ...children: NodeInput[]) => node("detailPanel", { id, presentation: "detailPanel", ...props }, ...children).region("side"),
  inlinePanel: (id: string, props: AdminJsonObject = {}, ...children: NodeInput[]) => node("inlinePanel", { id, presentation: "inlinePanel", ...props }, ...children),
};

export const resource = {
  page: (id: string, title = id) => new AdminPageBuilder(id, title, "resource"),
  list: (id: string, props: AdminJsonObject = {}, ...children: NodeInput[]) => node("resourceList", { id, ...props }, ...children),
  table: (id: string, columns: AdminJsonObject[], rows: AdminJsonObject[], props: AdminJsonObject = {}) => node("resourceTable", { id, columns, rows, ...props }).id(id),
  row: (id: string, props: AdminJsonObject = {}) => node("resourceRow", { id, ...props }).id(id),
  detail: (id: string, props: AdminJsonObject = {}, ...children: NodeInput[]) => node("resourceDetail", { id, ...props }, ...children),
};

export const field = {
  text: (name: string, props: AdminJsonObject = {}) => node("textField", { name, ...props }),
  textarea: (name: string, props: AdminJsonObject = {}) => node("textareaField", { name, ...props }),
  money: (name: string, props: AdminJsonObject = {}) => node("moneyField", { name, ...props }),
  duration: (name: string, props: AdminJsonObject = {}) => node("durationField", { name, ...props }),
  date: (name: string, props: AdminJsonObject = {}) => node("dateField", { name, ...props }),
  time: (name: string, props: AdminJsonObject = {}) => node("timeField", { name, ...props }),
  select: (name: string, options: AdminJsonObject[], props: AdminJsonObject = {}) => node("selectField", { name, options, ...props }),
  switch: (name: string, props: AdminJsonObject = {}) => node("switchField", { name, ...props }),
  image: (name: string, props: AdminJsonObject = {}) => node("imageField", { name, ...props }),
};

export const view = {
  list: (id: string, label = id, props: AdminJsonObject = {}) => ({ id, label, kind: "list", ...props }),
  calendar: (id: string, label = id, props: AdminJsonObject = {}) => ({ id, label, kind: "calendar", ...props }),
};

File: pkg/admindsl/builder.go
package admindsl

import (
	"encoding/json"
	"fmt"
)

// PageBuilder is the authoritative host-side fluent builder for Admin DSL pages.
// Goja modules should expose this builder (or a thin wrapper around it) rather
// than reimplementing Admin DSL validity rules in JavaScript.
type PageBuilder struct {
	page Page
	err  error
}

func PageResource(id, title string) *PageBuilder  { return NewPage(id, title, ShellResource) }
func PageAdmin(id, title string) *PageBuilder     { return NewPage(id, title, ShellAdmin) }
func PageDashboard(id, title string) *PageBuilder { return NewPage(id, title, ShellDashboard) }
func PageCalendar(id, title string) *PageBuilder  { return NewPage(id, title, ShellCalendar) }
func PageSettings(id, title string) *PageBuilder  { return NewPage(id, title, ShellSettings) }

func NewPage(id, title string, shell ShellKind) *PageBuilder {
	return &PageBuilder{page: Page{SchemaVersion: 1, ID: id, Title: title, Shell: Shell{Kind: shell}, Nodes: []Node{}}}
}

func (b *PageBuilder) Description(description string) *PageBuilder {
	b.page.Description = description
	return b
}

func (b *PageBuilder) Shell(kind ShellKind, props JSONObject) *PageBuilder {
	b.page.Shell = Shell{Kind: kind, Props: props}
	return b
}

func (b *PageBuilder) Meta(meta PageMeta) *PageBuilder {
	b.page.Meta = meta
	return b
}

func (b *PageBuilder) Toolbar(actions ...*ActionBuilder) *PageBuilder {
	b.page.Nodes = append([]Node{Toolbar(actions...).Build()}, b.page.Nodes...)
	return b
}

func (b *PageBuilder) Content(nodes ...*NodeBuilder) *PageBuilder {
	b.page.Nodes = buildNodes(nodes...)
	return b
}

func (b *PageBuilder) Add(nodes ...*NodeBuilder) *PageBuilder {
	b.page.Nodes = append(b.page.Nodes, buildNodes(nodes...)...)
	return b
}

func (b *PageBuilder) Modals(nodes ...*NodeBuilder) *PageBuilder {
	b.page.Modals = buildNodes(nodes...)
	return b
}

func (b *PageBuilder) Drawers(nodes ...*NodeBuilder) *PageBuilder {
	b.page.Drawers = buildNodes(nodes...)
	return b
}

func (b *PageBuilder) Build() (Page, error) {
	if b.err != nil {
		return Page{}, b.err
	}
	if err := ValidatePage(b.page); err != nil {
		return Page{}, err
	}
	return clonePage(b.page)
}

func (b *PageBuilder) MustBuild() Page {
	page, err := b.Build()
	if err != nil {
		panic(err)
	}
	return page
}

func (b *PageBuilder) JSON() ([]byte, error) {
	page, err := b.Build()
	if err != nil {
		return nil, err
	}
	return json.Marshal(page)
}

type NodeBuilder struct {
	node Node
}

func NodeOf(kind NodeKind, props JSONObject, children ...*NodeBuilder) *NodeBuilder {
	return (&NodeBuilder{node: Node{Kind: kind, Props: props}}).Children(children...)
}

func (b *NodeBuilder) ID(id string) *NodeBuilder {
	b.ensureMeta().ID = id
	return b
}

func (b *NodeBuilder) Name(name string) *NodeBuilder {
	b.ensureMeta().Name = name
	return b
}

func (b *NodeBuilder) Region(region Region) *NodeBuilder {
	b.ensureMeta().Region = region
	return b
}

func (b *NodeBuilder) DataSection(section string) *NodeBuilder {
	b.ensureMeta().DataSection = section
	return b
}

func (b *NodeBuilder) DataPart(part string) *NodeBuilder {
	b.ensureMeta().DataPart = part
	return b
}

func (b *NodeBuilder) Note(note string) *NodeBuilder {
	b.ensureMeta().Note = note
	return b
}

func (b *NodeBuilder) Props(props JSONObject) *NodeBuilder {
	if b.node.Props == nil {
		b.node.Props = JSONObject{}
	}
	for k, v := range props {
		b.node.Props[k] = v
	}
	return b
}

func (b *NodeBuilder) Children(children ...*NodeBuilder) *NodeBuilder {
	b.node.Children = buildNodes(children...)
	return b
}

func (b *NodeBuilder) Add(children ...*NodeBuilder) *NodeBuilder {
	b.node.Children = append(b.node.Children, buildNodes(children...)...)
	return b
}

func (b *NodeBuilder) Action(slot string, action *ActionBuilder) *NodeBuilder {
	if slot == "" {
		panic("admin dsl action slot must not be empty")
	}
	if b.node.Props == nil {
		b.node.Props = JSONObject{}
	}
	actions, _ := b.node.Props["actions"].(JSONObject)
	if actions == nil {
		actions = JSONObject{}
	}
	actions[slot] = action.Build()
	b.node.Props["actions"] = actions
	return b
}

func (b *NodeBuilder) Actions(actions ...*ActionBuilder) *NodeBuilder {
	values := make([]JSONValue, 0, len(actions))
	for _, action := range actions {
		values = append(values, action.Build())
	}
	if b.node.Props == nil {
		b.node.Props = JSONObject{}
	}
	b.node.Props["actions"] = values
	return b
}

func (b *NodeBuilder) Query(query *QueryBuilder) *NodeBuilder {
	if b.node.Props == nil {
		b.node.Props = JSONObject{}
	}
	b.node.Props["query"] = query.Build()
	return b
}

func (b *NodeBuilder) State(state string) *NodeBuilder {
	return b.Props(JSONObject{"state": state})
}

func (b *NodeBuilder) Values(values JSONObject) *NodeBuilder {
	return b.Props(JSONObject{"values": values})
}

func (b *NodeBuilder) Errors(errors JSONObject) *NodeBuilder {
	return b.Props(JSONObject{"errors": errors})
}

func (b *NodeBuilder) Submit(action *ActionBuilder) *NodeBuilder {
	return b.Action("submit", action)
}

func (b *NodeBuilder) Cancel(action *ActionBuilder) *NodeBuilder {
	return b.Action("cancel", action)
}

func (b *NodeBuilder) Dirty(dirty bool) *NodeBuilder {
	return b.Props(JSONObject{"dirty": dirty})
}

func (b *NodeBuilder) Pending(pending bool) *NodeBuilder {
	return b.Props(JSONObject{"pending": pending})
}

func (b *NodeBuilder) LayoutPolicy(policy JSONObject) *NodeBuilder {
	return b.Props(JSONObject{"layoutPolicy": policy})
}

func (b *NodeBuilder) Adaptive(views JSONObject) *NodeBuilder {
	return b.Props(JSONObject{"adaptive": views})
}

func (b *NodeBuilder) Build() Node {
	return b.node
}

func (b *NodeBuilder) ensureMeta() *NodeMeta {
	if b.node.Meta == nil {
		b.node.Meta = &NodeMeta{}
	}
	return b.node.Meta
}

type ActionBuilder struct {
	ref ActionRef
}

func Action(kind ActionType, target string, label string) *ActionBuilder {
	return &ActionBuilder{ref: ActionRef{Type: kind, Target: target, Label: label}}
}

func Open(target, label string) *ActionBuilder     { return Action(ActionOpen, target, label) }
func Close(target string) *ActionBuilder           { return Action(ActionClose, target, "") }
func Navigate(target, label string) *ActionBuilder { return Action(ActionNavigate, target, label) }
func Mutation(target, label string) *ActionBuilder { return Action(ActionMutation, target, label) }
func Confirm(target, label string) *ActionBuilder  { return Action(ActionConfirm, target, label) }
func Refresh(target, label string) *ActionBuilder  { return Action(ActionRefresh, target, label) }
func Upload(target, label string) *ActionBuilder   { return Action(ActionUpload, target, label) }

func (b *ActionBuilder) Payload(payload JSONValue) *ActionBuilder {
	b.ref.Payload = payload
	return b
}

func (b *ActionBuilder) Options(options JSONObject) *ActionBuilder {
	if b.ref.Options == nil {
		b.ref.Options = JSONObject{}
	}
	for k, v := range options {
		b.ref.Options[k] = v
	}
	return b
}

func (b *ActionBuilder) Intent(intent ActionIntent) *ActionBuilder {
	b.ref.Intent = intent
	return b
}

func (b *ActionBuilder) Priority(priority ActionPriority) *ActionBuilder {
	b.ref.Priority = priority
	return b
}

func (b *ActionBuilder) Placement(placement ActionPlacement) *ActionBuilder {
	b.ref.Placement = placement
	return b
}

func (b *ActionBuilder) Presentation(presentation string) *ActionBuilder {
	b.ref.Presentation = presentation
	return b
}

func (b *ActionBuilder) ConfirmRequired() *ActionBuilder {
	b.ref.RequiresConfirmation = true
	return b
}

func (b *ActionBuilder) Disabled(disabled bool) *ActionBuilder {
	b.ref.Disabled = disabled
	return b
}

func (b *ActionBuilder) Loading(loading bool) *ActionBuilder {
	b.ref.Loading = loading
	return b
}

func (b *ActionBuilder) AccessibilityLabel(label string) *ActionBuilder {
	b.ref.AccessibilityLabel = label
	return b
}

func (b *ActionBuilder) Build() ActionRef {
	return b.ref
}

func Primary(target, label string) *ActionBuilder {
	return Mutation(target, label).Intent(IntentPrimary).Priority(PriorityPrimary)
}

func Secondary(target, label string) *ActionBuilder {
	return Mutation(target, label).Intent(IntentNeutral).Priority(PrioritySecondary)
}

func Danger(target, label string) *ActionBuilder {
	return Mutation(target, label).Intent(IntentDanger).Priority(PrioritySecondary).ConfirmRequired()
}

type QueryBuilder struct {
	ref QueryRef
}

func Query(id string, params JSONObject) *QueryBuilder {
	return &QueryBuilder{ref: QueryRef{ID: id, Params: params}}
}

func (b *QueryBuilder) Params(params JSONObject) *QueryBuilder {
	b.ref.Params = params
	return b
}

func (b *QueryBuilder) Build() QueryRef {
	return QueryRef{ID: b.ref.ID, Params: cloneObject(b.ref.Params)}
}

func Section(title string, props JSONObject, children ...*NodeBuilder) *NodeBuilder {
	return NodeOf(NodeSection, merge(JSONObject{"title": title}, props), children...)
}

func Toolbar(actions ...*ActionBuilder) *NodeBuilder {
	return NodeOf(NodeToolbar, JSONObject{}).Actions(actions...)
}

func Panel(title string, props JSONObject, children ...*NodeBuilder) *NodeBuilder {
	return NodeOf(NodePanel, merge(JSONObject{"title": title}, props), children...)
}

func CardGrid(props JSONObject, children ...*NodeBuilder) *NodeBuilder {
	return NodeOf(NodeCardGrid, props, children...)
}

func Tabs(id string, props JSONObject) *NodeBuilder {
	return NodeOf(NodeTabs, merge(JSONObject{"id": id}, props)).ID(id)
}

func EditableList(id string, props JSONObject) *NodeBuilder {
	return NodeOf(NodeEditableList, merge(JSONObject{"id": id}, props)).ID(id)
}

func MonthAvailabilityGrid(id string, props JSONObject) *NodeBuilder {
	return NodeOf(NodeMonthAvailability, merge(JSONObject{"id": id}, props)).ID(id)
}

func PreviewFrame(id string, props JSONObject) *NodeBuilder {
	return NodeOf(NodePreviewFrame, merge(JSONObject{"id": id}, props)).ID(id)
}

func DiffView(id string, props JSONObject) *NodeBuilder {
	return NodeOf(NodeDiffView, merge(JSONObject{"id": id}, props)).ID(id)
}

func Metric(label string, value JSONValue, props JSONObject) *NodeBuilder {
	return NodeOf(NodeMetricCard, merge(JSONObject{"label": label, "value": value}, props))
}

func SummaryCard(title string, props JSONObject, children ...*NodeBuilder) *NodeBuilder {
	return NodeOf(NodeSummaryCard, merge(JSONObject{"title": title}, props), children...)
}

func EmptyState(title string, props JSONObject) *NodeBuilder {
	return NodeOf(NodeEmptyState, merge(JSONObject{"title": title}, props))
}

func ImageGallery(id string, props JSONObject) *NodeBuilder {
	return NodeOf(NodeImageGallery, merge(JSONObject{"id": id}, props)).ID(id)
}

func Markdown(markdown string, props JSONObject) *NodeBuilder {
	return NodeOf(NodeMarkdownBlock, merge(JSONObject{"markdown": markdown}, props))
}

func ResourceList(id string, props JSONObject, children ...*NodeBuilder) *NodeBuilder {
	return NodeOf(NodeResourceList, merge(JSONObject{"id": id}, props), children...)
}

func ResourceTable(id string, props JSONObject) *NodeBuilder {
	return NodeOf(NodeResourceTable, merge(JSONObject{"id": id}, props)).ID(id)
}

func FilterBar(id string, props JSONObject) *NodeBuilder {
	return NodeOf(NodeFilterBar, merge(JSONObject{"id": id}, props)).ID(id)
}

func SearchBox(id string, props JSONObject) *NodeBuilder {
	return NodeOf(NodeSearchBox, merge(JSONObject{"id": id}, props)).ID(id)
}

func ResourceRow(id string, props JSONObject) *NodeBuilder {
	return NodeOf(NodeResourceRow, merge(JSONObject{"id": id}, props)).ID(id)
}

func ResourceDetail(id string, props JSONObject, children ...*NodeBuilder) *NodeBuilder {
	return NodeOf(NodeResourceDetail, merge(JSONObject{"id": id}, props), children...)
}

func Form(id string, props JSONObject, children ...*NodeBuilder) *NodeBuilder {
	return NodeOf(NodeForm, merge(JSONObject{"id": id}, props), children...)
}

func FieldGroup(title string, children ...*NodeBuilder) *NodeBuilder {
	return NodeOf(NodeFieldGroup, JSONObject{"title": title}, children...)
}

func TextField(name string, props JSONObject) *NodeBuilder {
	return NodeOf(NodeTextField, merge(JSONObject{"name": name}, props))
}

func Drawer(id string, props JSONObject, children ...*NodeBuilder) *NodeBuilder {
	return NodeOf(NodeDrawer, merge(JSONObject{"id": id, "presentation": "drawer"}, props), children...).Region(RegionDrawer)
}

func Modal(id string, props JSONObject, children ...*NodeBuilder) *NodeBuilder {
	return NodeOf(NodeModal, merge(JSONObject{"id": id, "presentation": "modal"}, props), children...).Region(RegionModal)
}

func Sheet(id string, props JSONObject, children ...*NodeBuilder) *NodeBuilder {
	return NodeOf(NodeSheet, merge(JSONObject{"id": id, "presentation": "sheet"}, props), children...).Region(RegionDrawer)
}

func DetailPanel(id string, props JSONObject, children ...*NodeBuilder) *NodeBuilder {
	return NodeOf(NodeDetailPanel, merge(JSONObject{"id": id, "presentation": "detailPanel"}, props), children...).Region(RegionSide)
}

func InlinePanel(id string, props JSONObject, children ...*NodeBuilder) *NodeBuilder {
	return NodeOf(NodeInlinePanel, merge(JSONObject{"id": id, "presentation": "inlinePanel"}, props), children...)
}

func ConfirmDialog(id string, props JSONObject) *NodeBuilder {
	return NodeOf(NodeConfirmDialog, merge(JSONObject{"id": id, "presentation": "confirm"}, props)).Region(RegionModal)
}

func buildNodes(builders ...*NodeBuilder) []Node {
	nodes := make([]Node, 0, len(builders))
	for _, builder := range builders {
		if builder == nil {
			panic("admin dsl node builder must not be nil")
		}
		nodes = append(nodes, builder.Build())
	}
	return nodes
}

func merge(base JSONObject, extra JSONObject) JSONObject {
	out := JSONObject{}
	for k, v := range base {
		out[k] = v
	}
	for k, v := range extra {
		out[k] = v
	}
	return out
}

func clonePage(page Page) (Page, error) {
	data, err := json.Marshal(page)
	if err != nil {
		return Page{}, fmt.Errorf("marshal admin dsl page: %w", err)
	}
	var out Page
	if err := json.Unmarshal(data, &out); err != nil {
		return Page{}, fmt.Errorf("unmarshal admin dsl page clone: %w", err)
	}
	return out, nil
}

func cloneNode(node Node) Node {
	data, err := json.Marshal(node)
	if err != nil {
		panic(fmt.Errorf("marshal admin dsl node: %w", err))
	}
	var out Node
	if err := json.Unmarshal(data, &out); err != nil {
		panic(fmt.Errorf("unmarshal admin dsl node clone: %w", err))
	}
	return out
}

func cloneAction(action ActionRef) ActionRef {
	data, err := json.Marshal(action)
	if err != nil {
		panic(fmt.Errorf("marshal admin dsl action: %w", err))
	}
	var out ActionRef
	if err := json.Unmarshal(data, &out); err != nil {
		panic(fmt.Errorf("unmarshal admin dsl action clone: %w", err))
	}
	return out
}

func cloneObject(obj JSONObject) JSONObject {
	if obj == nil {
		return nil
	}
	data, err := json.Marshal(obj)
	if err != nil {
		panic(fmt.Errorf("marshal admin dsl object: %w", err))
	}
	var out JSONObject
	if err := json.Unmarshal(data, &out); err != nil {
		panic(fmt.Errorf("unmarshal admin dsl object clone: %w", err))
	}
	return out
}
